contracts/0.8.9/WithdrawalQueue.sol

```diff
@@ -0,0 +1,444 @@
+    function claimWithdrawals(ClaimWithdrawalInput[] calldata _claimWithdrawalInputs) external {
+        for (uint256 i = 0; i < _claimWithdrawalInputs.length; ++i) {
+            claimWithdrawal(_claimWithdrawalInputs[i].requestId, _claimWithdrawalInputs[i].hint);
+        }
+    }
+
+    /// @notice Finds the list of hints for the given `_requestIds` searching among the discounts with indices
+    ///  in the range  `[_firstIndex, _lastIndex]`
+    /// @param _requestIds ids of the requests sorted in the ascending order to get hints for
+    /// @param _firstIndex left boundary of the search range
+    /// @param _lastIndex right boundary of the search range
+    /// @return hintIds the hints for `claimWithdrawal` to find the discount for the passed request ids
+    function findClaimHints(uint256[] calldata _requestIds, uint256 _firstIndex, uint256 _lastIndex)
+        public
+        view
+        returns (uint256[] memory hintIds)
+    {
+        hintIds = new uint256[](_requestIds.length);
+        uint256 prevRequestId = 0;
+        for (uint256 i = 0; i < _requestIds.length; ++i) {
+            if (_requestIds[i] < prevRequestId) revert RequestIdsNotSorted();
+            hintIds[i] = findClaimHint(_requestIds[i], _firstIndex, _lastIndex);
+            _firstIndex = hintIds[i];
+            prevRequestId = _requestIds[i];
+        }
+    }
+
+    /// @notice Finds the list of hints for the given `_requestIds` searching among the discounts with indices
+    ///  in the range `[0, lastDiscountIndex]`
+    /// @dev WARNING! OOG is possible if used onchain.
+    ///  See `findClaimHints(uint256[] calldata _requestIds, uint256 _firstIndex, uint256 _lastIndex)` for onchain use
+    /// @param _requestIds ids of the requests sorted in the ascending order to get hints for
+    function findClaimHintsUnbounded(uint256[] calldata _requestIds) public view returns (uint256[] memory hintIds) {
+        return findClaimHints(_requestIds, 0, lastDiscountIndex);
+    }
+
+    /**
+     * @notice Finalize requests from last finalized one up to `_lastRequestIdToFinalize`
+     * @dev ether to finalize all the requests should be calculated using `finalizationBatch()` and sent along
+     *
+     * @param _lastRequestIdToFinalize request index in the queue that will be last finalized request in a batch
+     */
+    function finalize(uint256 _lastRequestIdToFinalize) external payable whenResumed onlyRole(FINALIZE_ROLE) {
+        _finalize(_lastRequestIdToFinalize, msg.value);
+    }
+
+    /**
+     * @notice Update bunker mode state
+     * @dev should be called by oracle
+     *
+     * NB: timestamp should correspond to the previous oracle report
+     *
+     * @param _isBunkerModeNow oracle report
+     * @param _previousOracleReportTimestamp timestamp of the previous oracle report
+     */
+    function updateBunkerMode(
+        bool _isBunkerModeNow,
+        uint256 _previousOracleReportTimestamp
```

Comment:
Let's rename `uint256 _bunkerModeSinceTimestamp` because this contract shouldn't know about oracle prev rebase